src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -990,247 +1067,423 @@ RouterSelectTask(Query *originalQuery, Query *query)
 	task->upsertQuery = upsertQuery;
 	task->requiresMasterEvaluation = false;
 
+	*placementList = workerList;
+
 	return task;
 }
 
 
 /*
- * RouterQueryJob creates a Job for the specified query to execute the
- * provided single shard select task.
+ * TargetShardIntervalsForSelect performs shard pruning for all referenced relations
+ * in the query and returns list of shards per relation. Shard pruning is done based
+ * on provided restriction context per relation. The function bails out and returns NULL
+ * if any of the relations has no active shards. It also records pruned shard intervals
+ * in relation restriction context to be used later on.
  */
-static Job *
-RouterQueryJob(Query *query, Task *task)
+static List *
+TargetShardIntervalsForSelect(Query *query,
+							  RelationRestrictionContext *restrictionContext)
 {
-	Job *job = NULL;
-	List *taskList = NIL;
-	TaskType taskType = task->taskType;
+	List *prunedRelationShardList = NIL;
+	ListCell *restrictionCell = NULL;
 
-	/*
-	 * We send modify task to the first replica, otherwise we choose the target shard
-	 * according to task assignment policy.
-	 */
-	if (taskType == MODIFY_TASK)
-	{
-		taskList = FirstReplicaAssignTaskList(list_make1(task));
-	}
-	else
+	Assert(query->commandType == CMD_SELECT);
+	Assert(restrictionContext != NULL);
+
+	foreach(restrictionCell, restrictionContext->relationRestrictionList)
 	{
-		taskList = AssignAnchorShardTaskList(list_make1(task));
-	}
+		RelationRestriction *relationRestriction =
+			(RelationRestriction *) lfirst(restrictionCell);
+		Oid relationId = relationRestriction->relationId;
+		Index tableId = relationRestriction->index;
+		DistTableCacheEntry *cacheEntry = DistributedTableCacheEntry(relationId);
+		int shardCount = cacheEntry->shardIntervalArrayLength;
+		List *baseRestrictionList = relationRestriction->relOptInfo->baserestrictinfo;
+		List *restrictClauseList = extract_actual_clauses(baseRestrictionList, false);
+		List *shardIntervalList = NIL;
+		List *prunedShardList = NIL;
+		int shardIndex = 0;
 
-	job = CitusMakeNode(Job);
-	job->dependedJobList = NIL;
-	job->jobId = INVALID_JOB_ID;
-	job->subqueryPushdown = false;
-	job->jobQuery = query;
-	job->taskList = taskList;
+		relationRestriction->prunedShardIntervalList = NIL;
 
-	return job;
+		for (shardIndex = 0; shardIndex < shardCount; shardIndex++)
+		{
+			ShardInterval *shardInterval =
+				cacheEntry->sortedShardIntervalArray[shardIndex];
+			shardIntervalList = lappend(shardIntervalList, shardInterval);
+		}
+
+
+		if (shardCount > 0)
+		{
+			prunedShardList = PruneShardList(relationId, tableId,
+											 restrictClauseList,
+											 shardIntervalList);
+		}
+
+		relationRestriction->prunedShardIntervalList = prunedShardList;
+		prunedRelationShardList = lappend(prunedRelationShardList, prunedShardList);
```

Comment:
One thing I'm a bit worried about, is whether we need a "quick exit" here, if shardCount is > 1 after pruning. For a complicated query we could spend a long time uselessly pruning, long after we've already essentially figured out there's no way this can be router executed.